services/blockchain-indexer/shared/utils/snapshot.js

```diff
@@ -0,0 +1,184 @@
+const util = require('util');
+const crypto = require('crypto');
+const config = require('../../config');
+const execInShell = util.promisify(require('child_process').exec);
+const { requestConnector } = require('./request');
+
+const logger = Logger();
+
+let snapshotFilePath = './data/service-snapshot.sql';
+
+const getHTTPProtocolByURL = (url) => url.startsWith('https') ? https : http;
+
+const checkCommandAvailability = async () => {
+	const { stdout: mysqlAvailable } = await execInShell('which mysql').catch(() => ({}));
+	if (!mysqlAvailable) throw new NotFoundException('mysql command is unavailable in PATH.');
+};
+
+const calculateSHA256 = async (file) => new Promise((resolve, reject) => {
+	const hash = crypto.createHash('sha256');
+	const stream = fs.createReadStream(file);
+
+	stream.on('data', (data) => hash.update(data));
+	stream.on('end', () => resolve(hash.digest('hex')));
+	stream.on('error', (error) => reject(error));
+});
+
+const downloadUnzipAndVerifyChecksum = async (fileUrl, filePath) => {
+	const checksumUrl = fileUrl.replace('.gz', '.SHA256');
+
+	return new Promise((resolve, reject) => {
+		// Download the checksum file
+		logger.info('Attempting to download the snapshot checksum.');
+		getHTTPProtocolByURL(checksumUrl).get(checksumUrl, (response) => {
+			if (response.statusCode === 200) {
+				let checksumData = '';
+
+				response.on('data', (chunk) => {
+					checksumData += chunk;
+				});
+
+				response.on('end', () => {
+					// Extract the SHA256 hash from the downloaded data
+					const checksum = checksumData.trim().split(' ')[0];
+
+					// Download and unzip the file
+					logger.info('Attempting to download the snapshot file.');
+					getHTTPProtocolByURL(fileUrl).get(fileUrl, (res) => {
+						if (res.statusCode === 200) {
+							const unzip = zlib.createUnzip();
+							const writeFile = fs.createWriteStream(filePath);
+
+							res.pipe(unzip).pipe(writeFile);
+
+							res.on('error', (err) => {
+								reject(new Error(err));
+							});
+
+							writeFile.on('finish', () => {
+								// calculate hash from file.
+								calculateSHA256(filePath).then((calculatedChecksum) => {
+									if (calculatedChecksum === checksum) {
+										resolve();
+									} else {
+										reject(new Error('Checksum verification failed.'));
+									}
+								}).catch(err => {
+									reject(err);
+								});
+							});
+
+							writeFile.on('error', (err) => {
+								reject(err);
+							});
+						} else {
+							const errMessage = `Download failed with HTTP status code: ${res.statusCode} (${res.statusMessage})`;
+							console.error(errMessage);
+							if (res.statusCode === 404) {
+								reject(new Error(`NotFoundException: ${errMessage}`));
+							} else {
+								reject(new Error(errMessage));
+							}
+						}
+					}).on('error', (err) => {
+						reject(new Error(err));
+					});
+				});
+			} else {
+				reject(new Error('Failed to download checksum file.'));
```

Comment:
Could we please also log the response code, statusMessage and any other necessary information?
```suggestion
				reject(new Error('Failed to download the checksum file.'));
```